trackers_integration/issuetracker/trac.py

```diff
@@ -0,0 +1,202 @@
+        # in Trac cannot determine permissions
+        url = f"{self.__base_url}/{project}/login"
+        resp = session.get(url, timeout=30, auth=self.__auth)
+        if resp.status_code != http.HTTPStatus.OK:
+            return {"error": f"{resp.status_code}: {resp.reason}"}
+        # now invoke RPC method on Trac server
+        url = f"{self.__base_url}/{project}/ticketrpc"
+        req = {
+            "jsonrpc": "2.0",
+            "method": method,
+            "params": args,
+            "id": str(time.time_ns()),
+        }
+        resp = session.post(
+            url, timeout=30, headers=self.__headers, auth=self.__auth, json=req
+        )
+        rc = resp.status_code
+        return (
+            resp.json()
+            if rc == http.HTTPStatus.OK
+            else {"error": f"{rc}: {resp.reason}"}
+        )
+
+
+class Trac(IssueTrackerType):
+    """
+    .. versionadded:: 15.1-Enterprise
+
+    Support for `Trac <https://trac.edgewall.org/>`_ - open source
+    issue tracking system, version 1.6 and above.
+    Trac server must have plugin trac-ticketrpc installed (https://pypi.org/project/trac-ticketrpc)
+    Project name in Trac must match product name in Kiwi TCMS.
+
+    .. warning::
+
+        Make sure that this package is installed inside Kiwi TCMS and that
+        ``EXTERNAL_BUG_TRACKERS`` setting contains a dotted path reference to
+        this class! When using *Kiwi TCMS Enterprise* this is configured
+        automatically.
+
+    **Authentication**:
+
+    :base_url: URL to Trac instance - e.g. https://trac.myserver.local/
+    :api_username: Trac Username - needs Trac permissions TICKET_CREATE, TICKET_APPEND, TICKET_VIEW
+    :api_password: Trac Password
+    """
+
+    def _rpc_connection(self):
+        user, password = self.rpc_credentials
+        return TracAPI(self.bug_system.base_url, user, password)
+
+    def is_adding_testcase_to_issue_disabled(self):
+        user, password = self.rpc_credentials
+        return not (self.bug_system.base_url and user and password)
+
+    def _report_issue(self, execution, _user):
+        """
+        Create Trac ticket.
+        :param execution: test execution data
+        :param _user: current TCMS user
+        :return: Response from Trac server, issue ID
+        """
+        product = execution.build.version.product.name
+        try:
+            version = execution.build.version.value
+            summary = f"Failed test: {execution.case.summary}"
+            description = execution.case.get_text_with_version(
+                execution.case_text_version
+            )
+            ticket = {
+                "type": "defect",
+                "priority": "major",
+                "summary": summary,
+                "description": description,
+                "project": product,
+                "version": version,
+                "component": product,
+            }
+            result = self.rpc.invoke_method("ticket.create", ticket)
```

Comment:
IMO lines 117-127 should be split into a create_ticket() method on the TracAPI class and then the same can be used both here as well as in tests by passing the correct dict to it.